pkg/setup/binary_hyperkube.go

```diff
@@ -9,43 +9,62 @@ import (
 	"fmt"
 	"os"
 	"os/exec"
+	"text/template"
+
+	"github.com/Masterminds/semver"
 
 	"github.com/golang/glog"
 )
 
 func (e *Environment) extractHyperkube() error {
 	glog.V(2).Infof("Extracting %s", e.binaryHyperkube.archivePath)
-	b, err := exec.Command("tar", "-C", e.binABSPath, "-xzvf", e.binaryHyperkube.archivePath, "--strip-components=3", "kubernetes/server/bin/hyperkube").CombinedOutput()
+	b, err := exec.Command("tar", "-C", e.binABSPath, "-xzvf", e.binaryHyperkube.archivePath, "--strip-components=3").CombinedOutput()
 	output := string(b)
 	if err != nil {
 		glog.Errorf("Cannot untar %s, %s: %v", e.binaryHyperkube.archivePath, output, err)
 		_ = e.binaryHyperkube.removeArchive()
 		return err
 	}
-	_, err = os.Stat(e.binaryHyperkube.binaryABSPath)
+
+	// After 1.17 hyperkube binary has been removed from Kubernetes
+	// It's replaced by a shell script forwarding calls to each binary
+	hyperkubeMissingCheck, err := semver.NewConstraint(">=1.17.0")
 	if err != nil {
-		glog.Errorf("Unexpected error: %v after untar %s", err, output)
 		return err
 	}
-	glog.V(2).Infof("Successfully untar %s: %s", e.binaryHyperkube.archivePath, output)
-	wd, err := os.Getwd()
-	if err != nil {
-		glog.Errorf("Cannot get wd: %v", err)
-		return err
+
+	if hyperkubeMissingCheck.Check(e.kubeVersion) {
```

Comment:
Why not directly doing the needed configuration here instead of going through a templated bash script ?
So we could, for example do the extraction of the binaries in the version before 1.17 or generate the symlinks.

Reply:
I'm not sure I fully understand how I can replace the shell script, maybe I missed something.

I believe the only thing we could do is change the code everywhere for all versions to use binaries separately, be it real binaries or symlink for each binary to hyperkube binary for versions < 1.17 (though I don't see what that would bring, as we would have already changed the code to call binaries separately). As I wanted to minimize the changes to support 1.17+ versions (cf. what do we do with p8s in the future?), I added the shell script.

Note that this shell script (without template) is also bundled and used in the hyperkube docker image we deploy inside kube.

Reply:
If I remember correctly, it's possible to generate symlink to `hyperkube` and then have the `kube-apiserver` for example, directly called from the binary path, instead of having `${PATH}/hyperkube apiserver`.
That could avoid us this additional moving templated bash script layer by solving this with a proper configuration migration.

If you think this is too big to achieve, feel free to merge as is, once the tests are green.

Reply:
Correct, it's currently used for kubectl (and thus, I've added support for it in the bash script). But in our case, the issue is the opposite, we have all binaries (for all versions now, since I unpack the full tarball) and we need to dispatch calls that are made to `hyperkube <binary>` to `<binary>`, thus 
the shell script.

Otherwise we do full migration and we don't need any symlink or bash script (cleanest option).

I'll check with the team their opinion on this.

Reply:
We will currently continue with the minimum effort to support newer Kube versions. An RFC is open internally with comparison of alternatives, but it requires to properly define what we expect from the e2e CI Tests.